kernel/src/schema.rs

```diff
@@ -21,6 +21,14 @@ pub enum MetadataValue {
     Number(i32),
     String(String),
     Boolean(bool),
+    // to support things like icebergv2 compat, where we can have fields like:
+    // "parquet.field.nested.ids": {
+    //    "col1.element": 100,
+    //    "col1.element.element": 101
+    // }
+    // TODO: Perhaps make this more restrictive, the PROTOCOL seems to indicate that this should
+    // always be a `String->Int` map
+    Object(serde_json::Map<String, serde_json::Value>),
```

Comment:
So then IIUC this is letting column/structfield metadata also include arbitrary string->json? Is there any API here other than the ability to just ask for the metadata back?

Not that i'm against it, just thinking of some implications: this makes serde_json::Map/Value part of the public API right?

We could also take the extreme approach and do `Json(serde_json::Value)`? or do the most specific of Map<Int, String>. or keep the in-between currently in the PR.

This comment is mostly questions/ideas I've thought of - I think most would be reasonable here?

Reply:
It does indeed make it part of the public API. I don't think this is exactly correct, but I think there's some work to be done to spec out in the protocol exactly what can go here.

This basically allows:
```json
{
  "key": [number],
  "key": [string],
  "key": [bool],
  "key": {
      [pretty much any object]
   }
}
```

Allowing `serde_json::Map<String, serde_json::Value>` is basically equivalent to "anything in a json object" because json objects _are_ just `string -> Value` maps. So what we're disallowing here is:
```json
{
  "key": [1,2,3], // array
  "key": null
}
```

and that's it. I can't find use for that in the spec so I slightly prefer the more restrictive form. We can relax this later without breaking things, but it's harder to go from more loose to tighter.

Reply:
are we disallowing array/null? Isn't there `serde_json::Value::Null` and `serde_json::Value::Array`?

Reply:
it just means the top level of enum there has to be an object, what is contained in the map can technically be anything `serde_json::Value` can take

Reply:
> are we disallowing array/null? Isn't there `serde_json::Value::Null` and `serde_json::Value::Array`?

We are only disallowing it at the top level. Inside the `Object` variant you could have anything, but at the top we only deser into the types I listed above.

Reply:
The [spec](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#struct-field) only says `metadata` is a "JSON map" and AFAIK delta-spark (spark, really) represents it as a String -> Any map (see [Metadata.scala](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/types/Metadata.scala#L43)). So technically even Null and Array values would be allowed.

Reply:
Fair enough, I've updated this to now just have an `Other` variant that will catch anything we don't understand, plus a comment explaining why that's okay.

Reply:
i misunderstood yay thanks lgtm :)